plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -319,21 +333,29 @@ getCardanoTxRedeemers = onCardanoTx txRedeemers (const Map.empty) -- TODO: To im
     --         let d' = Datum $ CardanoAPI.fromCardanoScriptData d in Just (datumHash d', d')
     --     fromCardanoTxOutDatum (C.TxOutDatumInline _ d) =
     --         let d' = Datum $ CardanoAPI.fromCardanoScriptData d in Just (datumHash d', d')
+getCardanoTxRedeemers :: CardanoTx -> Map V1.ScriptPurpose V1.Redeemer
+getCardanoTxRedeemers = onCardanoTx txRedeemers (const Map.empty)
 
+-- Defined here as uses `txId`.
 instance Pretty Tx where
-    pretty tx =
+    pretty tx@(Tx _txInputs _txReferenceInputs _txCollateral _txOutputs _txMint _txFee
+                 _txValidRange _txMintingScripts _txWithdrawals _txCertificates
+                 _txSignatures _txScripts _txData _txMetadata) =
         let lines' =
-                [ hang 2 (vsep ("inputs:" : fmap pretty (txInputs tx)))
-                , hang 2 (vsep ("reference inputs:" : fmap pretty (txReferenceInputs tx)))
-                , hang 2 (vsep ("collateral inputs:" : fmap pretty (txCollateral tx)))
-                , hang 2 (vsep ("outputs:" : fmap pretty (txOutputs tx)))
-                , "mint:" <+> pretty (txMint tx)
-                , "fee:" <+> pretty (txFee tx)
-                , hang 2 (vsep ("mps:": fmap pretty (Map.toList $ txMintScripts tx)))
-                , hang 2 (vsep ("signatures:": fmap (pretty . fst) (Map.toList $ txSignatures tx)))
-                , "validity range:" <+> viaShow (txValidRange tx)
-                , hang 2 (vsep ("redeemers:": fmap (pretty . snd) (Map.toList $ txRedeemers tx) ))
-                , hang 2 (vsep ("data:": fmap (pretty . snd) (Map.toList $ txData tx) ))
+                [ hang 2 (vsep ("inputs:" : fmap pretty _txInputs))
+                , hang 2 (vsep ("reference inputs:" : fmap pretty _txReferenceInputs))
+                , hang 2 (vsep ("collateral inputs:" : fmap pretty _txCollateral))
+                , hang 2 (vsep ("outputs:" : fmap pretty _txOutputs))
+                , "mint:" <+> pretty _txMint
+                , "fee:" <+> pretty _txFee
+                , hang 2 (vsep ("mps:": fmap pretty (Map.assocs _txMintingScripts)))
+                , hang 2 (vsep ("signatures:": fmap (pretty . fst) (Map.toList _txSignatures)))
+                , "validity range:" <+> viaShow _txValidRange
+                , hang 2 (vsep ("data:": fmap (pretty . snd) (Map.toList _txData)))
+                , hang 2 (vsep ("attached scripts:": fmap pretty (Map.keys _txScripts)))
```

Comment:
I think we should only show these 4 lines when they are not empty, this adds quite a bit of noise to the logs.

Reply:
metadata line as well, right?

Reply:
Fixed by "Pretty print if non-empty."